harp/reader.py

```diff
@@ -44,13 +45,37 @@ def __init__(
         self.read = read
 
 
+class RegisterMap(UserDict[str, RegisterReader]):
+    _address_map: dict[int, str]
+
+    def __init__(self, registers: dict[str, RegisterReader]) -> None:
+        super().__init__(registers)
+        self._address_map = {self[key].register.address: key for key in self.keys()}
+
+    def __getitem__(self, __key: Union[str, int]) -> RegisterReader:
+        if isinstance(__key, int):
+            super().__getitem__(self._address_map[__key])
+        elif isinstance(__key, str):
+            return super().__getitem__(__key)
+        else:
+            raise TypeError(f"key must be int or str, not {type(__key)}")
+
+    def __getattr__(self, __name: str) -> RegisterReader:
+        return self[__name]
+
+
 class DeviceReader:
     device: Model
-    registers: dict[str, RegisterReader]
+    registers: RegisterMap
 
-    def __init__(self, device: Model, registers: dict[str, RegisterReader]) -> None:
+    def __init__(self, device: Model, registers: Union[RegisterMap, dict[str, RegisterReader]]) -> None:
         self.device = device
-        self.registers = registers
+        if isinstance(registers, RegisterMap):
+            self.registers = registers
+        elif isinstance(registers, dict):
+            self.registers = RegisterMap(registers)
+        else:
+            raise TypeError(f"registers must be dict or RegisterMap, not {type(registers)}")
```

Comment:
As we are not enforcing type validation for other attributes in the package, I would rather not add an exception here and simply leave a direct assignment in the `else` clause.

There is also an advantage that any `dict`-like object could be passed as a register map here, including other `UserDict` objects (interestingly `UserDict` is not always a subclass of `dict` depending on your python version!).

```suggestion
        else:
            self.registers = RegisterMap(registers)
```